### ISSUE TITLE
Performance degradation when migrating from gRpc.Core to gRpc.AspNetCore  

### ISSUE THREAD

**Issue by YaronDror23** [grpc/grpc-dotnet#1531]
We are currently using gRpc.Core in our solution and we would like to migrate to the new libraries (gRpc.AspNetCore).
Since we are using .net framework 4.6.2 we followed the recommendation from this article (https://docs.microsoft.com/en-us/aspnet/core/grpc/netstandard?view=aspnetcore-5.0) and used grcp-web (gRcp.AspNetCore.Web).
We wanted to compare the performance of our current setup against the new one and we saw a significant performance degradation.
The test we ran was a simple single threaded client sending non stop update stream requests 

This is the API :
service StreamingApi {
  rpc UpdateStream(stream UpdateRequest) returns (UpdateResponse);
}

We ran 4 scenarios :
1. gRpc.Core Client -> gRpc.Core Server (processed 30,000 requests per sec with average latency of 1 sec)
2. gRpc.Net.Client.Web Client -> gRpc.AspNetCore.Web Server (processed 300 requests per sec with average latency of 1 sec)
3. gRpc.Core Client -> gRpc.AspNetCore Server (managed to process 40,000 requests per sec with average latency of 1 sec)
4. grpc.Net.Client Client -> gRpc.AspNetCore Server (managed to process 1200 requests per sec with average latency of 1 sec)

The results took us by surprise (to say the least .. )
It looks as if the new server is performing better than the old one but the new client performers much worse.

Is it common knowledge that grpc-web performance is much worse than grpc ? (I guess it has to do with http1 vs http2 )
Is there any other solution for .net framework users ? (we need to use bidirectional streaming)
Are there any tweaks we can do to improve the new client performance ?     
Is it a reasonable solution to migrate our server to gRpc.AspNetCore and keep working with gRpc.Core client ? 

Thanks

**Comment by JamesNK** [comment]
Grpc.Net.Client support on .NET Framework is new and hasn't been benchmarked or optimized. Right now we're focused on getting it working.

Could you share the code that creates the GrpcChannel?

I see your test method is using client streaming. Client streaming support is brand new in .NET 6 and needs a modern version of Windows. What Windows version are you using?

Also, gRPC-Web doesn't officially support client streaming. It might work using Grpc.Net.Client, but it hasn't been tested closely.

**Comment by YaronDror23** [comment]
Hey James

Thanks for the quick reply.

Regarding your questions :

1. Q : Could you share the code that creates the GrpcChannel?
A : I followed your instructions from this article (https://docs.microsoft.com/en-us/aspnet/core/tutorials/grpc/grpc-start?view=aspnetcore-6.0&tabs=visual-studio) 
This is the code 

    var channel = GrpcChannel.ForAddress("http://localhost:5000");
    var client = new StreamingApiClient(channel);

    ...        

    client.Update(new UpdateRequest());

2. Q : I see your test method is using client streaming. Client streaming support is brand new in .NET 6 and needs a modern version of Windows. What Windows version are you using?
A : The test I performed was with a client written in .Net 6 using gRcp.Net.Client with a .Net 6 server app using gRcp.AspNetCore server. So there are no windows version requirements.  

I saw that you wrote here (https://github.com/grpc/grpc-dotnet/blob/master/doc/implementation_comparison.md) 
"Note that there's ongoing work to add grpc-dotnet client support on legacy .NET Framework."

When will this be available ? Is there a pre prelease version we can start using ?

Is there any other solutions for .net framework users ? grcp-web is not a valid option for us, throughput wise 

Thanks in advance

**Comment by JamesNK** [comment]
I thought you were using .NET Framework. You also mentioned using gRPC-Web. The way you're creating a client won't work with either of them. WinHttpHandler is required.
```cs
var channel = GrpcChannel.ForAddress("http://localhost:5000");
var client = new StreamingApiClient(channel);
```
I'm confused.

**Comment by YaronDror23** [comment]
Hey James

Let me clarify.

We ran 4 scenarios :

1. gRpc.Core Client on a .net framework 4.6.2 app using grcp mode -> gRpc.Core Server on a .net core 3.1 app using grcp mode
         (processed 30,000 requests per sec with average latency of 1 sec)
2. gRpc.Net.Client.Web Client on a .net framework 4.8 app using grcp-web mode-> gRpc.AspNetCore.Web Server on a .Net 6 app using grcp-web mode  (processed 300 requests per sec with average latency of 1 sec)
3. gRpc.Core Client on a .net framework 4.6.2 app using grcp mode -> gRpc.AspNetCore Server on a .Net 6 app using grcp mode (managed to process 40,000 requests per sec with average latency of 1 sec)
4. grpc.Net.Client Client on a .Net 6 app using grcp mode -> gRpc.AspNetCore Server on a .Net 6 app using grcp mode (managed to process 1200 requests per sec with average latency of 1 sec)


The results of the grcp-web mode were bad but I contributed them to the lack of support in Http/2, what really surprised me
were the results of the 4'th test which in my opinion points to a performance issue in the gRpc.NetClient lib.

I have uploaded all my code to https://github.com/jivygroup/gRpcPerformancePoc
You are welcome to dive in.

The big question from our side is how to move forward to use gRpc.NetClient in our .framework 4.8 app ??

**Comment by JamesNK** [comment]
This is your problem: https://github.com/jivygroup/gRpcPerformancePoc/blob/27dc0fde6f499c8a14a07bd04ce78c325ebf3dd2/PerformanceTests-gRcp.AspNetCore/Net6ClientGrpcMode/LightStreamerProxyClient.cs#L31

**Comment by YaronDror23** [comment]
Hey James

Thanks for your feedback.

The line you mentioned did cause an issue and when I switched it to client streaming the performance was much better.
It is still less than half the throughput we were able to generate with the old libraries, so I think that this is something worth looking into, but this not the issue we are facing, we performed this test (the 4'th one) just as sanity.

The problem we are facing is the lack of support for .net framework.

With the gRpc.Core libraries we were able to have a .net framework client working with a .net core server with very good performance. Now we are facing a limitation that if we want to use gRcp.AspNetCore in .net framework we can:
1. Be constrained to a Windows build and not use client streaming - no an option for us, our use case includes client streaming and we don't host our applications so we have very basic control over our machines.
2. Use grpc-web - that also doesn't support client streaming 

So basically we have no option to migrate to gRpc.AspNetCore as we would like to do per your recommendation. 

Can you recommend what can we do to work with grcp with a .net framework client and a .net core server ?

**Comment by JamesNK** [comment]
Your client choices are:
1. Keep using Grpc.Core
2. Update app to not use .NET Framework and use Grpc.Net.Client
3. Use a modern version of Windows that supports client-side streaming with `WinHttpHandler`. I think the only Windows version that supports it is the latest Win11 insiders dev build.

**Comment by YaronDror23** [comment]
Hey James

1. We will probably choose to work with gRcp.Core client in our .net framework apps. Do you see anything we need to consider for that or any limitations we will have in the future? 
2. We also have a few .Net core clients which we would to work with gRcp.AspNetCore. Can you update on the effort on your side in the performance area?
3. I saw that here (https://github.com/grpc/grpc-dotnet/blob/master/doc/implementation_comparison.md)
"Note that there's ongoing work to add grpc-dotnet client support on legacy .NET Framework." Is this still relevant ?

Thanks

**Comment by JamesNK** [comment]
> We will probably choose to work with gRcp.Core client in our .net framework apps. Do you see anything we need to consider for that or any limitations we will have in the future?

Grpc.Core is being deprecated over the next couple of years.

> We also have a few .Net core clients which we would to work with gRcp.AspNetCore. Can you update on the effort on your side in the performance area?

Could you give me the performance numbers comparing Grpc.Net.Client with Grpc.Core? When I used your app I found that Grpc.Net.Client was faster.

> I saw that here (master/doc/implementation_comparison.md)
"Note that there's ongoing work to add grpc-dotnet client support on legacy .NET Framework." Is this still relevant ?

Yes. Waiting for Windows builds to become publically available.

**Comment by YaronDror23** [comment]
Hey James

Thanks for all you feedback, you have been very helpful.

Looking forward for updates on the new .net framework supported client.

Happy new year

**Comment by YaronDror23** [comment]
Hey James

One more thing I would like to clarify.

I your last response you said - "When I used your app I found that Grpc.Net.Client was faster."
But my results show otherwise. 

When I ran the gRpc.Core server (https://github.com/jivygroup/gRpcPerformancePoc/tree/main/PerformanceTests-gRcp.Core/NetCore3.1Server) with the gRpc.Core client (https://github.com/jivygroup/gRpcPerformancePoc/tree/main/PerformanceTests-gRcp.Core/NetFramework4.6.2Client)
The server was able to process ~4500 requests per second.
When I ran the gRpc.AspNetCore server (https://github.com/jivygroup/gRpcPerformancePoc/tree/main/PerformanceTests-gRcp.AspNetCore/Net6ServerGrpcMode) with the gRpc.AspNetCore client (https://github.com/jivygroup/gRpcPerformancePoc/tree/main/PerformanceTests-gRcp.AspNetCore/Net6ClientGrpcMode)
The server was able to process ~2000 requests per second.

Here is a screenshot of my results 

![gRpcCore](https://user-images.githubusercontent.com/56866577/148178378-c4b2e816-a8bf-476a-a2d7-ea2f5ed288ff.JPG)

Do have any explanation for this ?
Is it OS related ?

Thanks

**Comment by JamesNK** [comment]
I don't know why you are getting that result.

**Comment by yevgenyfst** [comment]
Hi James,
What are the arguments you thinking of?

**Comment by yevgenyfst** [comment]
Thank you for the double check. We will check in our code....

Lets close this issue  then ...
